src/gc/softwarewritewatch.cpp

```diff
@@ -0,0 +1,243 @@
+        {
+            if (GetDirtyFromBlock(
+                    currentBlock,
+                    firstPageAddressInCurrentBlock,
+                    tableRegionStart - blockStart,
+                    tableRegionEnd - fullBlockEnd,
+                    dirtyPages,
+                    &dirtyPageIndex,
+                    dirtyPageCount,
+                    clearDirty))
+            {
+                *dirtyPageCountRef = dirtyPageIndex;
+            }
+            break;
+        }
+
+        if (tableRegionStart != blockStart)
+        {
+            if (!GetDirtyFromBlock(
+                    currentBlock,
+                    firstPageAddressInCurrentBlock,
+                    tableRegionStart - blockStart,
+                    sizeof(size_t),
+                    dirtyPages,
+                    &dirtyPageIndex,
+                    dirtyPageCount,
+                    clearDirty))
+            {
+                break;
+            }
+            currentBlock += sizeof(size_t);
+            firstPageAddressInCurrentBlock += sizeof(size_t) * OS_PAGE_SIZE;
+        }
+
+        while (currentBlock < fullBlockEnd)
+        {
+            if (!GetDirtyFromBlock(
+                    currentBlock,
+                    firstPageAddressInCurrentBlock,
+                    0,
+                    sizeof(size_t),
+                    dirtyPages,
+                    &dirtyPageIndex,
+                    dirtyPageCount,
+                    clearDirty))
+            {
+                break;
+            }
+            currentBlock += sizeof(size_t);
+            firstPageAddressInCurrentBlock += sizeof(size_t) * OS_PAGE_SIZE;
+        }
+        if (currentBlock < fullBlockEnd)
+        {
+            break;
+        }
+
+        if (tableRegionEnd != fullBlockEnd &&
+            !GetDirtyFromBlock(
+                currentBlock,
+                firstPageAddressInCurrentBlock,
+                0,
+                tableRegionEnd - fullBlockEnd,
+                dirtyPages,
+                &dirtyPageIndex,
+                dirtyPageCount,
+                clearDirty))
+        {
+            break;
+        }
+
+        *dirtyPageCountRef = dirtyPageIndex;
+    } while (false);
+
+    if (!isRuntimeSuspended && clearDirty && dirtyPageIndex != 0)
+    {
+        // When dirtying a page, the dirty state of the page is first checked to see if the page is already dirty. If already
+        // dirty, the write to mark it as dirty is skipped. So, when the dirty state of a page is cleared, we need to make sure
+        // the cleared state is visible to other threads that may dirty the page, before marking through objects in the page, so
+        // that the GC will not miss marking through dirtied objects in the page. Issue a memory barrier on all active threads
+        // of the process now.
+        MemoryBarrier(); // flush writes from this thread first to guarantee ordering
+        FlushProcessWriteBuffers();
```

Comment:
Please use GCToOSInterface::FlushProcessWriteBuffers to keep the GC code OS independent